Task 7

vladimir-arsenov opened the pull request.

vladimir-arsenov committed:
Task 7

vladimir-arsenov committed:
Add tests

vladimir-arsenov committed:
Add currency conversion test

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/client/CachingConfig.java:
```diff
@@ -0,0 +1,29 @@
+        return new ConcurrentMapCacheManager("currenciesRates");
+    }
+
+    @CacheEvict(value = "currenciesRates", allEntries = true)
```
так не делают, интервал очищения можно задать в пропертях, иначе на каждый кеш в приложении у тебя будет куча таких методов

а в проперти мало того, что из коробки кеш предоставляет функционал еще и будет все в одном месте и будешь видеть валюты через час очищаются, чеки через 4 и тд

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/client/CentralBankClient.java:
```diff
@@ -0,0 +1,61 @@
+        return parseRates(restTemplate.getForEntity(currencyRatesUrl, String.class).getBody());
+    }
+
+    @SneakyThrows
```
в пет проекте эта анноташка база, но тут лучше бы обработать исключения)

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/controller/CurrencyController.java:
```diff
@@ -0,0 +1,35 @@
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/currencies")
+public class CurrencyController {
```
за минимализм плюс балл, но нет спеки для свагера

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/dto/CurrencyConversionRequest.java:
```diff
@@ -0,0 +1,21 @@
+
+    @NotNull(message = "amount is mandatory")
+    @Min(value = 0, message = "amount can't be less than zero")
+    private Float amount;
```
BigDecimal

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/dto/CurrencyConversionResponse.java:
```diff
@@ -0,0 +1,12 @@
+public class CurrencyConversionResponse {
+    private String fromCurrency;
+    private String toCurrency;
+    private Float convertedAmount;
```
BigDecimal

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,56 @@
+
+@Slf4j
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
👍

lParzivall commented on currency-service/src/main/java/org/example/currencyservice/service/CurrencyService.java:
```diff
@@ -0,0 +1,62 @@
+        var to = rates.get(conversionRequest.getToCurrency());
+        var amount = conversionRequest.getAmount();
+
+        var result = from.getRate() * amount / to.getRate();
```
округление

vladimir-arsenov committed:
add swagger specification;
remove @SneakyThrows;
fix caching configuration;
fix: use BigDecimal to work with currencies;
